ORC-743: Added conversion of SArg into filters to take advantage of t…

pavibhai opened the pull request:
### What changes were proposed in this pull request?
Added conversion of SArg into filters to take advantage of the LazyIO introduced by ORC-742
    * Created Vector filters for leaf, And, Or, Batch
    * Code generation for data type and operator filters
    * Test code generation for data type and operator filters

### Why are the changes needed?
With ORC-742, FOLLOW (non-filter) columns are evaluated lazily. Spark and Hive are already passing down SearchArguments to ORC. This change allows the conversion of the SearchArguments to a filter to gain the improvements of ORC-742.

### How was this patch tested?
* Unit tests were added

dongjoon-hyun commented on java/core/pom.xml:
```diff
@@ -151,12 +160,40 @@
-        <groupId>org.codehaus.mojo</groupId>
-        <artifactId>build-helper-maven-plugin</artifactId>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-shade-plugin</artifactId>
```
Could you add this after `build-helper-maven-plugin`?

pavibhai replied:
done

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -23,7 +23,8 @@
 import java.time.format.DateTimeFormatter;
-import org.apache.commons.lang.ArrayUtils;
+
+import org.apache.commons.lang3.ArrayUtils;
```
Just a question. Is it required? Or, can we do this first as a small PR?

pavibhai replied:
Cleaned up

dongjoon-hyun commented on java/core/src/java/org/apache/orc/util/CuckooSetBytes.java:
```diff
@@ -0,0 +1,443 @@
+ * For details on the algorithm, see R. Pagh and F. F. Rodler, "Cuckoo Hashing,"
+ * Elsevier Science preprint, Dec. 2003. http://www.itu.dk/people/pagh/papers/cuckoo-jour.pdf.
+ *
+ * Copied from CuckooSetBytes@Apache Hive project for convenience
```
Could you add a hyperlink here?

pavibhai replied:
Will do

pavibhai replied:
The links is pretty long "https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/CuckooSetBytes.java" which causes now line length violation, do you want to fragment this link in the comment?

dongjoon-hyun commented on java/core/src/java/org/apache/orc/util/CuckooSetBytes.java:
```diff
@@ -0,0 +1,443 @@
+ *
+ * Copied from CuckooSetBytes@Apache Hive project for convenience
+ */
+public class CuckooSetBytes {
```
And, let's proceed this separately. Then, I can help you merge this first safely.

pavibhai replied:
But this will not be used if we add this first. Is that ok?

dongjoon-hyun replied:
Yes, it's okay. Since this PR is the mother PR, we can spin off some small PRs from this.

dongjoon-hyun commented on java/core/src/test/org/apache/orc/impl/TestSchemaEvolution.java:
```diff
@@ -763,8 +763,6 @@ public void testTimestampToDecimal64Evolution() throws Exception {
   @Test
   public void testTimestampToStringEvolution() throws Exception {
-    testFilePath = new Path(workDir, "TestSchemaEvolution." +
-                                         testCaseName.getMethodName() + ".orc");
```
Is there a reason why you remove this in your new feature PR, @pavibhai ?

pavibhai replied:
Oh yeah this is a cleanup, I can raise a separate one for this.

dongjoon-hyun replied:
Yep. Please make it separately. Then, we can quickly merge it.

dongjoon-hyun commented on java/core/src/test/org/apache/orc/util/CuckooSetBytesTest.java:
```diff
@@ -0,0 +1,124 @@
+
+import static org.junit.Assert.*;
+
+public class CuckooSetBytesTest {
```
Could you add a reference if this is copied together?

pavibhai replied:
Will do

dongjoon-hyun commented on java/gen/src/findbugs/exclude.xml:
```diff
@@ -0,0 +1,17 @@
+-->
+<FindBugsFilter>
+
+</FindBugsFilter>
```
Just a question. Do we need this empty file?

pavibhai replied:
yeah this was needed otherwise the analyze was throwing error

dongjoon-hyun replied:
Could you share the exceptions you hit?

pavibhai replied:
Try running 
```bash
mvn clean package -Panalyze
```

Then you get the error
```text
[ERROR] Could not find resource '--SNIP--orc/java/gen/src/findbugs/exclude.xml'. -> [Help 1]
```

dongjoon-hyun commented on java/mapreduce/src/java/org/apache/orc/mapred/OrcInputFormat.java:
```diff
@@ -153,8 +153,10 @@ public static void setSearchArgument(Configuration conf,
+    //Mapreduce supports selected vector
+    Reader.Options options =  buildOptions(conf, file, split.getStart(), split.getLength())
+      .allowSelected(true);
+    return new OrcMapredRecordReader<>(file,options);
```
nit. `file,options` -> `file, options`

pavibhai replied:
done

dongjoon-hyun reviewed (commented):
Thank you for making a PR, @pavibhai . I left a few comments. In general, let's split the PR into smaller ones in order to make it easier for review and merging.

pavibhai commented:
Thanks @dongjoon-hyun I have addressed some of the review comments.

Requesting @pgaref and @omalley to also go through the PR when they get time.

omalley commented on java/core/pom.xml:
```diff
@@ -37,6 +37,11 @@
       <groupId>org.apache.orc</groupId>
       <artifactId>orc-shims</artifactId>
     </dependency>
+    <dependency>
```
I pushed the results of removing the code generation in a [fork](https://github.com/omalley/orc/tree/orc-743) of this branch.

pavibhai replied:
I agree that generation is not having a lot of value given the separation of LeafFilter which also accounts for NOT operations now.

I have removed this.

omalley commented on java/core/src/gen/filters/string_eq.txt:
```diff
@@ -0,0 +1,44 @@
+import java.nio.charset.StandardCharsets;
+
+// This is generated from string_eq.txt
+public class <ClassName> extends LeafFilter {
```
All of the leaf classes should be package local.

pavibhai replied:
Done

omalley commented on java/core/src/gen/filters/type_in.txt:
```diff
@@ -0,0 +1,45 @@
+
+// This is generated from type_in.txt
+public class <ClassName> extends LeafFilter {
+  public final <LeafType>[] inValues;
```
We should have all of the fields as private.

pavibhai replied:
done

omalley commented on java/core/src/java/org/apache/orc/filter/impl/OrFilter.java:
```diff
@@ -0,0 +1,65 @@
+    this.filters = filters;
+  }
+
+  public static void merge(Selected src, Selected tgt) {
```
This should be moved to Selected and renamed to unionDistinct.

pavibhai replied:
moved and named it as unionDisjoint

omalley commented on java/core/src/java/org/apache/orc/filter/VectorFilter.java:
```diff
@@ -0,0 +1,41 @@
+   * @param selIn  The current selection
+   * @param selOut The result selection
+   */
+  void filter(OrcFilterContext fc, Selected bound, Selected selIn, Selected selOut);
```
I'd propose that we should join bound and selIn to be the rows that should be checked. The documentation should make it clear that bound should not be modified. Furthermore, we document that items in selOut that are not in bound must be retained. The selOut vector must be sorted.

pavibhai replied:
Makes sense.

* Removed selIn
* Updated the documentation

omalley commented on java/core/src/java/org/apache/orc/filter/impl/LeafFilter.java:
```diff
@@ -0,0 +1,97 @@
+import org.apache.orc.OrcFilterContext;
+import org.apache.orc.filter.VectorFilter;
+
+public abstract class LeafFilter implements VectorFilter {
```
I'd suggest putting a boolean in to LeafFilter that is whether the filter is negated. It can be used on the calls to accept at very low cost.

pavibhai replied:
Makes sense, incorporating this.

omalley reviewed (changes requested):
I think that most of these classes should move into org.apache.orc.impl.filter. The leaf classes and the factory class should be in org.apache.orc.impl.filter.leaf.

omalley commented on java/core/src/java/org/apache/orc/filter/impl/IsNullFilter.java:
```diff
@@ -0,0 +1,78 @@
+    // If the vector does not have nulls then none of them are selected and nothing to do
+    if (!noNulls) {
+      if (v.isRepeating) {
+        // If the repeating vector is null then set all as selected.
```
This assumes that the v.isNull[0] is true, which isn't valid.

pavibhai replied:
Fixed

omalley commented on java/core/src/java/org/apache/orc/filter/impl/IsNullFilter.java:
```diff
@@ -0,0 +1,78 @@
+        // If the repeating vector is null then set all as selected.
+        for (int i = 0; i < bound.selSize; i++) {
+          // Identify the rowIdx from the selected vector
+          rowIdx = bound.sel[i];
```
I'd suggest making a method in Selected (selectAll?) that creates the identity list of all of the rows up to N.

pavibhai replied:
Good idea. Adding this.

pavibhai commented on java/core/pom.xml:
```diff
@@ -114,6 +119,10 @@
     <dependency>
       <groupId>net.bytebuddy</groupId>
       <artifactId>byte-buddy</artifactId>
+    </dependency>
```
Removed

dongjoon-hyun commented:
Now, `main` branch has hive-storage-api 2.8.1.

pavibhai commented:
> Now, `main` branch has hive-storage-api 2.8.1.

Thanks @dongjoon-hyun will incorporate the changes of `using compactExpression`

pavibhai commented:
@omalley Thanks for taking a detailed look into this. I have taken the branch you shared and did the following on top of the same. You will find separate commit for these changes.

* Added the filter tests
* Removed selIn as a parameter for the filter
* Addressed review comments from Owen
* Switched to compactExpression on the SearchArgument
* Nested the filter classes in core
* Removed JUnit dependency

dongjoon-hyun commented on java/core/src/java/org/apache/orc/util/CuckooSetBytes.java:
```diff
@@ -0,0 +1,462 @@
+ * A high-performance set implementation used to support fast set membership testing,
+ * using Cuckoo hashing. This is used to support fast tests of the form
+ *
+ *       column IN ( <list-of-values )
```
Unfortunately, this broke our `Javadoc` generation. Here is the fix.
- https://github.com/apache/orc/pull/867

pavibhai replied:
@dongjoon-hyun Thanks for taking care of this, appreciate that.
